pika/adapters/select_connection.py

```diff
@@ -632,22 +632,10 @@ def _dispatch_fd_events(self, fd_event_map):
     def _get_interrupt_pair():
         """ Use a socketpair to be able to interrupt the ioloop if called
         from another thread. Socketpair() is not supported on some OS (Win)
-        so use a pair of simple UDP sockets instead. The sockets will be
+        so use a pair of simple TCP sockets instead. The sockets will be
         closed and garbage collected by python when the ioloop itself is.
         """
-        try:
-            read_sock, write_sock = socket.socketpair()
-
-        except AttributeError:
-            LOGGER.debug("Using custom socketpair for interrupt")
-            read_sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
-            read_sock.bind(('localhost', 0))
-            write_sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
-            write_sock.connect(read_sock.getsockname())
-
-        read_sock.setblocking(0)
-        write_sock.setblocking(0)
```

Comment:
@lukebakken - WARNING The original code returned a non-blocking socket pair, but the pair is blocking after this PR. The change from non-blocking to blocking can lead to a deadlock. The implementation was designed for non-blocking model - see https://github.com/pika/pika/blob/2b145f4e40e4fa2dda3b5e94dba8a86fbaf83319/pika/adapters/select_connection.py#L539 and https://github.com/pika/pika/blob/2b145f4e40e4fa2dda3b5e94dba8a86fbaf83319/pika/adapters/select_connection.py#L651.

A simple/pathological case that would reproduce the deadlock is calling `stop` enough times from code executing within the event loop itself to fill the socket's transmit buffer.

A more complex - albeit still possibly pathological - case is where a background thread or many background threads call `stop`, while the event loop code is blocked waiting for something from the thread that happens to make the `stop` call that would exceed the `_w_interrupt` socket's transmit buffer.

Is there any reason for removing the logic that made the sockets non-blocking?

Reply:
Nope, I just missed it in the change. Addressing it now.